src/Chapter3.hs

```diff
@@ -452,14 +475,40 @@ concept of product types and sum types is called __Algebraic Data Type__. They
 allow you to model your domain precisely, make illegal states unrepresentable
 and provide more flexibility when working with data types.
 -}
+data Hero 
+    = DeadWarrior 
+    | Warrior
+      { warriorHealth :: Int
+      , warriorAttack :: Int
+      , warriorGold :: Int
+      } deriving (Eq, Show)
+
+fightToDeath :: Hero -> Monster -> Hero
+fightToDeath = heroTurn
+    where
+      heroTurn :: Hero -> Monster -> Hero
+      heroTurn DeadWarrior _ = DeadWarrior
+      heroTurn warrior@(Warrior heroHealth attack gold) monster@(Monster health _ enemyGold)
+          | heroHealth <= 0 = DeadWarrior
+          | healthDiff > 0 = warrior `enemyTurn` (monster {monsterHealth = healthDiff})
+          | otherwise = warrior {warriorGold = gold + enemyGold}
+          where healthDiff = health - attack
+
+      enemyTurn :: Hero -> Monster -> Hero
+      enemyTurn DeadWarrior _ = DeadWarrior
+      enemyTurn (Warrior 0 _ _) _ = DeadWarrior
+      enemyTurn warrior@(Warrior health _ _) monster@(Monster _ attack _)
+          | healthDiff > 0 = (warrior {warriorHealth = healthDiff}) `heroTurn` monster
+          | otherwise = DeadWarrior
+          where healthDiff = health - attack
```

Comment:
That is a very interesting solution! ⚔️ 

I feel like it could be simplified a bit.
I guess it is possible to represent one round with one function (where one round is knight + monster attack each other). And then this round function can be called recursively 🙂 I just see an opportunity to simplify a few cases and check all at once, however, your solution is already covers all cases 👍🏼 

And if you would change `Hero` type as I mentioned in the previous comment, I think, you could simplify some stuff in here too. As you noticed, you can't use `warrior*` functions in here, so no much benefits from the record in such representation for now.